File: frontend/src/types/oe.ts
// Sales Order Types
export interface SalesOrder {
  id: number;
  company_id: number;
  document_number: string;
  order_date: string;
  customer_id: number;
  reference?: string;
  sales_representative_id?: number;
  status: 'DRAFT' | 'CONFIRMED' | 'INVOICED' | 'CANCELLED' | 'Draft';
  currency_code?: string;
  exchange_rate?: number;
  subtotal?: number;
  tax_amount?: number;
  total_amount: number;
  notes?: string;
  ar_invoice_id?: number;
  
  // Related data
  customer_name?: string;
  sales_representative_name?: string;
  lines?: SalesOrderLine[];
}

export interface SalesOrderLine {
  id: number;
  sales_order_id: number;
  line_number: number;
  item_id: number;
  item_code?: string;
  item_description?: string;
  quantity: number;
  unit_price: number;
  discount_percentage: number;
  line_total: number;
  notes?: string;
}

export interface SalesOrderCreate {
  customer_id: number;
  order_date: string;
  reference?: string;
  sales_representative_id?: number;
  notes?: string;
  shipping_address?: any;
  billing_address?: any;
  lines: SalesOrderLineCreate[];
}

export interface SalesOrderLineCreate {
  item_id: number;
  description: string;
  quantity_ordered: number;
  unit_price: number;
  discount_percentage?: number;
  tax_type_id?: number;
}

// Purchase Order Types
export interface PurchaseOrder {
  id: number;
  company_id: number;
  document_number: string;
  order_number?: string; // Keep for backward compatibility
  order_date: string;
  supplier_id: number;
  supplier_reference?: string;
  status: 'DRAFT' | 'CONFIRMED' | 'RECEIVED' | 'INVOICED' | 'CANCELLED';
  currency_code: string;
  exchange_rate: number;
  subtotal: number;
  tax_amount: number;
  total_amount: number;
  notes?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  
  // Related data
  supplier_name?: string;
  lines?: PurchaseOrderLine[];
}

export interface PurchaseOrderLine {
  id: number;
  purchase_order_id: number;
  line_number: number;
  item_id: number;
  item_code?: string;
  item_description?: string;
  quantity: number;
  unit_price: number;
  discount_percentage: number;
  line_total: number;
  quantity_received: number;
  notes?: string;
}

export interface PurchaseOrderCreate {
  supplier_id: number;
  order_date: string;
  expected_delivery_date?: string;
  reference?: string;
  status?: string;
  total_amount: number;
  notes?: string;
  delivery_address_warehouse_id: number;
  lines: PurchaseOrderLineCreate[];
}

export interface PurchaseOrderLineCreate {
  item_id: number;
  description: string;
  quantity_ordered: number;
  unit_price: number;
  discount_percentage?: number;
  tax_type_id?: number;
  tax_amount?: number;
  line_total: number;
  notes?: string;
}

// Goods Received Voucher Types
export interface GoodsReceivedVoucher {
  id: number;
  company_id: number;
  document_number: string; // Backend uses document_number
  grv_number?: string; // Keep for backward compatibility
  grv_date: string;
  purchase_order_id?: number;
  supplier_id: number;
  supplier_delivery_note?: string;
  reference?: string; // Backend may use reference
  received_by?: string;
  status: 'DRAFT' | 'Open' | 'CONFIRMED' | 'INVOICED';
  total_value?: number;
  notes?: string;
  is_active?: boolean;
  ap_invoice_id?: number;
  created_at?: string;
  updated_at?: string;
  
  // Related data
  purchase_order_number?: string;
  supplier_name?: string;
  lines?: GoodsReceivedVoucherLine[];
}

export interface GoodsReceivedVoucherLine {
  id: number;
  grv_id: number;
  line_number: number;
  purchase_order_line_id?: number;
  item_id: number;
  item_code?: string;
  item_description?: string;
  quantity_ordered?: number;
  quantity_received: number;
  unit_cost: number;
  unit_price?: number; // Keep for backward compatibility
  line_total: number;
  notes?: string;
}

export interface GoodsReceivedVoucherCreate {
  supplier_id: number;
  purchase_order_id?: number;
  grv_date?: string;
  supplier_delivery_note?: string;
  reference?: string;
  received_by?: string;
  notes?: string;
  lines: GoodsReceivedVoucherLineCreate[];
}

export interface GoodsReceivedVoucherLineCreate {
  item_id: number;
  description?: string;
  quantity_received: number;
  unit_price?: number;
  unit_cost?: number;
  line_total?: number;
  purchase_order_line_id?: number;
  notes?: string;
}

// Order Defaults Types
export interface OrderDefaults {
  id: number;
  company_id: number;
  default_currency_code: string;
  default_payment_terms?: string;
  default_sales_representative_id?: number;
  default_warehouse_id?: number;
  auto_generate_order_numbers: boolean;
  sales_order_number_prefix?: string;
  purchase_order_number_prefix?: string;
  grv_number_prefix?: string;
  require_approval_for_orders: boolean;
  order_approval_limit?: number;
  created_at: string;
  updated_at: string;
  
  // Related data
  default_sales_representative_name?: string;
  default_warehouse_name?: string;
  // Order status defaults
  default_so_status: 'Draft' | 'Open' | 'PartiallyInvoiced' | 'Invoiced' | 'Closed' | 'Cancelled';
  default_po_status: 'Draft' | 'Open' | 'PartiallyReceived' | 'Received' | 'Closed' | 'Cancelled';
  default_grv_status: 'Open' | 'PartiallyInvoiced' | 'Invoiced' | 'Closed';
  // Next document numbers
  next_so_number: number;
  next_po_number: number;
  next_grv_number: number;
}

export interface OrderDefaultsUpdate {
  default_currency_code?: string;
  default_payment_terms?: string;
  default_sales_representative_id?: number;
  default_warehouse_id?: number;
  auto_generate_order_numbers?: boolean;
  sales_order_number_prefix?: string;
  purchase_order_number_prefix?: string;
  grv_number_prefix?: string;
  require_approval_for_orders?: boolean;
  order_approval_limit?: number;
  default_so_status?: 'Draft' | 'Open' | 'PartiallyInvoiced' | 'Invoiced' | 'Closed' | 'Cancelled';
  default_po_status?: 'Draft' | 'Open' | 'PartiallyReceived' | 'Received' | 'Closed' | 'Cancelled';
  default_grv_status?: 'Open' | 'PartiallyInvoiced' | 'Invoiced' | 'Closed';
}

// Re-export AR and AP Transaction types for conversions
export interface ARTransaction {
  id: number;
  company_id: number;
  transaction_type: string;
  transaction_number: string;
  transaction_date: string;
  customer_id: number;
  amount: number;
  status: string;
  reference?: string;
  // ... other AR transaction fields
}

export interface APTransaction {
  id: number;
  company_id: number;
  transaction_type: string;
  transaction_number: string;
  transaction_date: string;
  supplier_id: number;
  amount: number;
  status: string;
  reference?: string;
  // ... other AP transaction fields
}

File: backend/app/schemas/oe.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date
from decimal import Decimal

# Sales Order Schemas
class SalesOrderLineBase(BaseModel):
    item_id: int
    description: str
    quantity_ordered: Decimal
    unit_price: Decimal
    discount_percentage: Optional[Decimal] = 0
    tax_type_id: Optional[int] = None
    tax_amount: Optional[Decimal] = 0
    line_total: Decimal

class SalesOrderLineCreate(BaseModel):
    item_id: int
    description: str
    quantity_ordered: Decimal
    unit_price: Decimal
    discount_percentage: Optional[Decimal] = 0
    tax_type_id: Optional[int] = None

class SalesOrderLineUpdate(BaseModel):
    item_id: Optional[int] = None
    description: Optional[str] = None
    quantity_ordered: Optional[Decimal] = None
    unit_price: Optional[Decimal] = None
    discount_percentage: Optional[Decimal] = None
    tax_type_id: Optional[int] = None
    tax_amount: Optional[Decimal] = None
    line_total: Optional[Decimal] = None

class SalesOrderLine(SalesOrderLineBase):
    id: int
    sales_order_id: int
    quantity_invoiced: Decimal
    
    # Related item data
    item_code: Optional[str] = None
    item_description: Optional[str] = None

    class Config:
        from_attributes = True

class SalesOrderBase(BaseModel):
    customer_id: int
    order_date: date
    reference: Optional[str] = None
    status: str = "Draft"
    total_amount: Decimal
    notes: Optional[str] = None
    shipping_address: Optional[dict] = None
    billing_address: Optional[dict] = None
    sales_representative_id: Optional[int] = None

class SalesOrderCreate(BaseModel):
    customer_id: int
    order_date: date
    reference: Optional[str] = None
    notes: Optional[str] = None
    shipping_address: Optional[dict] = None
    billing_address: Optional[dict] = None
    sales_representative_id: Optional[int] = None
    lines: List[SalesOrderLineCreate] = []

class SalesOrderUpdate(BaseModel):
    customer_id: Optional[int] = None
    order_date: Optional[date] = None
    reference: Optional[str] = None
    status: Optional[str] = None
    total_amount: Optional[Decimal] = None
    notes: Optional[str] = None
    shipping_address: Optional[dict] = None
    billing_address: Optional[dict] = None
    sales_representative_id: Optional[int] = None
    lines: Optional[List[SalesOrderLineUpdate]] = None

class SalesOrder(SalesOrderBase):
    id: int
    company_id: int
    document_number: str
    ar_invoice_id: Optional[int] = None
    lines: List[SalesOrderLine] = []
    
    # Related data
    customer_name: Optional[str] = None
    sales_representative_name: Optional[str] = None
    currency_code: Optional[str] = "USD"
    exchange_rate: Optional[Decimal] = 1.0
    subtotal: Optional[Decimal] = None
    tax_amount: Optional[Decimal] = None

    class Config:
        from_attributes = True

# Purchase Order Schemas
class PurchaseOrderLineBase(BaseModel):
    item_id: int
    description: str
    quantity_ordered: Decimal
    unit_price: Decimal
    discount_percentage: Optional[Decimal] = 0
    tax_type_id: Optional[int] = None
    tax_amount: Optional[Decimal] = 0
    line_total: Decimal

class PurchaseOrderLineCreate(PurchaseOrderLineBase):
    pass

class PurchaseOrderLineUpdate(BaseModel):
    item_id: Optional[int] = None
    description: Optional[str] = None
    quantity_ordered: Optional[Decimal] = None
    unit_price: Optional[Decimal] = None
    discount_percentage: Optional[Decimal] = None
    tax_type_id: Optional[int] = None
    tax_amount: Optional[Decimal] = None
    line_total: Optional[Decimal] = None

class PurchaseOrderLine(PurchaseOrderLineBase):
    id: int
    purchase_order_id: int
    quantity_received: Decimal

    class Config:
        from_attributes = True

class PurchaseOrderBase(BaseModel):
    supplier_id: int
    order_date: date
    expected_delivery_date: Optional[date] = None
    reference: Optional[str] = None
    status: str = "Draft"
    total_amount: Decimal
    notes: Optional[str] = None
    delivery_address_warehouse_id: int

class PurchaseOrderCreate(PurchaseOrderBase):
    lines: List[PurchaseOrderLineCreate] = []

class PurchaseOrderUpdate(BaseModel):
    supplier_id: Optional[int] = None
    order_date: Optional[date] = None
    expected_delivery_date: Optional[date] = None
    reference: Optional[str] = None
    status: Optional[str] = None
    total_amount: Optional[Decimal] = None
    notes: Optional[str] = None
    delivery_address_warehouse_id: Optional[int] = None
    lines: Optional[List[PurchaseOrderLineUpdate]] = None

class PurchaseOrder(PurchaseOrderBase):
    id: int
    company_id: int
    document_number: str
    lines: List[PurchaseOrderLine] = []

    class Config:
        from_attributes = True

# Purchase Order Report Schema with additional fields
class PurchaseOrderReport(BaseModel):
    id: int
    order_number: str  # document_number
    supplier_id: int
    supplier_name: str
    order_date: date
    expected_delivery_date: Optional[date] = None
    reference: Optional[str] = None
    supplier_reference: Optional[str] = None  # from supplier.supplier_code or reference field
    status: str
    total_amount: Decimal
    currency_code: str = "USD"  # Default to USD for now
    notes: Optional[str] = None

    class Config:
        from_attributes = True

# Goods Received Voucher Schemas
class GoodsReceivedVoucherLineBase(BaseModel):
    item_id: int
    description: str
    quantity_received: Decimal
    unit_cost: Decimal
    line_total: Decimal
    purchase_order_line_id: Optional[int] = None

class GoodsReceivedVoucherLineCreate(GoodsReceivedVoucherLineBase):
    pass

class GoodsReceivedVoucherLineUpdate(BaseModel):
    item_id: Optional[int] = None
    description: Optional[str] = None
    quantity_received: Optional[Decimal] = None
    unit_cost: Optional[Decimal] = None
    line_total: Optional[Decimal] = None
    purchase_order_line_id: Optional[int] = None

class GoodsReceivedVoucherLine(GoodsReceivedVoucherLineBase):
    id: int
    grv_id: int
    
    # Related item data
    item_code: Optional[str] = None
    item_description: Optional[str] = None

    class Config:
        from_attributes = True

class GoodsReceivedVoucherBase(BaseModel):
    supplier_id: int
    grv_date: date
    reference: Optional[str] = None  # Supplier's Delivery Note No.
    status: str = "Open"
    notes: Optional[str] = None
    purchase_order_id: Optional[int] = None

class GoodsReceivedVoucherCreate(GoodsReceivedVoucherBase):
    lines: List[GoodsReceivedVoucherLineCreate] = []

class GoodsReceivedVoucherUpdate(BaseModel):
    supplier_id: Optional[int] = None
    grv_date: Optional[date] = None
    reference: Optional[str] = None
    status: Optional[str] = None
    notes: Optional[str] = None
    purchase_order_id: Optional[int] = None
    lines: Optional[List[GoodsReceivedVoucherLineUpdate]] = None

class GoodsReceivedVoucher(GoodsReceivedVoucherBase):
    id: int
    company_id: int
    document_number: str
    ap_invoice_id: Optional[int] = None
    lines: List[GoodsReceivedVoucherLine] = []
    
    # Related data
    supplier_name: Optional[str] = None
    purchase_order_number: Optional[str] = None
    total_value: Optional[Decimal] = None

    class Config:
        from_attributes = True

# Order Defaults Schemas
class OrderDefaultsBase(BaseModel):
    # Status defaults
    default_so_status: str = "Draft"
    default_po_status: str = "Draft"
    default_grv_status: str = "Open"
    
    # Document numbering
    next_so_number: int = 1
    next_po_number: int = 1
    next_grv_number: int = 1
    
    # General settings
    default_currency_code: str = "USD"
    default_payment_terms: Optional[str] = None
    default_sales_representative_id: Optional[int] = None
    default_warehouse_id: Optional[int] = None
    
    # Numbering settings
    auto_generate_order_numbers: bool = True
    sales_order_number_prefix: str = "SO-"
    purchase_order_number_prefix: str = "PO-"
    grv_number_prefix: str = "GRV-"
    
    # Approval settings
    require_approval_for_orders: bool = False
    order_approval_limit: Optional[float] = None

class OrderDefaultsCreate(OrderDefaultsBase):
    pass

class OrderDefaultsUpdate(BaseModel):
    # Status defaults
    default_so_status: Optional[str] = None
    default_po_status: Optional[str] = None
    default_grv_status: Optional[str] = None
    
    # Document numbering
    next_so_number: Optional[int] = None
    next_po_number: Optional[int] = None
    next_grv_number: Optional[int] = None
    
    # General settings
    default_currency_code: Optional[str] = None
    default_payment_terms: Optional[str] = None
    default_sales_representative_id: Optional[int] = None
    default_warehouse_id: Optional[int] = None
    
    # Numbering settings
    auto_generate_order_numbers: Optional[bool] = None
    sales_order_number_prefix: Optional[str] = None
    purchase_order_number_prefix: Optional[str] = None
    grv_number_prefix: Optional[str] = None
    
    # Approval settings
    require_approval_for_orders: Optional[bool] = None
    order_approval_limit: Optional[float] = None

class OrderDefaults(OrderDefaultsBase):
    id: int
    company_id: int

    class Config:
        from_attributes = True
